packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/usePrepareFileEntityUpload.ts

```diff
@@ -0,0 +1,113 @@
+import { SynapseClientError } from '@sage-bionetworks/synapse-client'
+import { useMutation, UseMutationOptions } from '@tanstack/react-query'
+import { useSynapseContext } from '../../context/index'
+import { getFileEntityIdWithSameName } from './getFileEntityIdWithSameName'
+import { useCreatePathsAndGetParentId } from './useCreatePathsAndGetParentId'
+
+export type FilePreparedForUpload = {
+  file: File
+  parentId: string
+  existingEntityId: string | null
+}
+
+export type PrepareDirsForUploadReturn = {
+  newFileEntities: FilePreparedForUpload[]
+  updatedFileEntities: FilePreparedForUpload[]
+}
+
+/**
+ * Mutation used to check and prepare the entity tree just before uploading a list of files.
+ *
+ * Given a list of files and a parent ID in which the files should be uploaded, the mutation will:
+ *
+ * 1. Create the necessary Folder entities in which the files to be uploaded (e.g. if the user uploaded a folder)
+ * 2. Check if any of the files to be uploaded already exist in the target parent folder
+ *
+ * The mutation will return two lists of files and their destination parentIds:
+ *    - newFileEntities: New files that do not have corresponding file entities in the target parent folder
+ *    - updatedFileEntities: Files that have corresponding file entities in the target parent folder. The user
+ *        should be prompted to accept or reject the creation of a new version of the file.
+ *
+ * In the future, this sequence could be amended to check if the storage location has enough space to accommodate all the
+ * new files, and return an error if not.
+ *
+ * @param options
+ */
+export function usePrepareFileEntityUpload(
```

Comment:
This wasn't captured in any design, but I'm slightly changing how we do this in SWC.

In SWC, each file is handled serially, so the process is something like (pseudocode)

```
for file in files:
  createDirectories(file)
  const fileEntityExists = checkIfFileEntityExists(file)
  if fileEntityExists:
    promptUserToUpdateFile()
  upload(file)
```

Where now it will be something like
```
const userPrompts = []
for file in files:
  createDirectories(file)
  const fileEntityExists = checkIfFileEntityExists(file)
  if fileEntityExists:
    userPrompts.push(file)

promptUserToUpdateFiles(userPrompts)
upload(files)
```

When we prompt the user "Do you want to create a new version of file xyz?" we can give the options "Yes", "Yes to all", "No", "Cancel uploads", so their upload won't be interrupted unless there's an error, and no uploads start until they have gone through each of these prompts.

WDYT about this? I think in the happy-path scenario this is much more user-friendly. A major downside to this approach is that a bunch of directories might be created unnecessarily

Reply:
There have been many requests to move away from serially processing, so that users do not have to babysit the upload (with this new approach, the user has the option to say "Yes to all" and walk away!).  I think the side effect creating the full directory tree up front is acceptable (given the user has not spent time uploading actual file content, which is the bulk of the operation, and they can always move the entire tree into the trash can if they don't want it).